kvserver: reduce replica mu contention in updateProposalQuotaRaftMuLocked

iskettaneh opened the pull request:
This PR refactors updateProposalQuotaRaftMuLocked() to reduce the replica mutex contention, especially in the normal case (QuotaPool is disabled, and the leadership is stable).

Changes done by this commit:

1. Create a short path for the case where we are not the leader.
2. Create a short path for the case where QuotaPool is not enabled.
We only take the write lock in cases where we already had proposal quota that needs to be release, or if the flow control level causes the replicaFlowControlIntegration.onRaftTicked to be a no-op.
3. Untangle the rest of the cases where the QuotaPool is enabled.


benchdiff results:

```
rm -f /tmp/*.pb.gz && env GODEBUG=runtimecontentionstacks=1 benchdiff -b --old e0393de6614ddad81a8209733945be46e967e993 -c 10 ./pkg/sql/tests -r BenchmarkParallelSysbench/SQL/3node/oltp_read_write --mutexprofile -d=3000x

name                                           old time/op    new time/op    delta
ParallelSysbench/SQL/3node/oltp_read_write-24    1.15ms ± 7%    1.15ms ± 4%   ~     (p=1.000 n=10+9)

name                                           old errs/op    new errs/op    delta
ParallelSysbench/SQL/3node/oltp_read_write-24      0.01 ±46%      0.01 ±32%   ~     (p=0.469 n=10+10)

name                                           old alloc/op   new alloc/op   delta
ParallelSysbench/SQL/3node/oltp_read_write-24    2.05MB ± 1%    2.06MB ± 1%   ~     (p=0.052 n=10+10)

name                                           old allocs/op  new allocs/op  delta
ParallelSysbench/SQL/3node/oltp_read_write-24     8.18k ± 1%     8.20k ± 1%   ~     (p=0.101 n=10+10)
```

Mutex contention graph after running the microbenchmark on my GCE worker:

```
rm -f /tmp/*.pb.gz && env GODEBUG=runtimecontentionstacks=1 go test ./pkg/sql/tests -run - -bench BenchmarkParallelSysbench/SQL/3node/oltp_read_write -v -test.benchtime=3000x -test.outputdir=/tmp -test.mutexprofile=mutex.pb.gz -test.mutexprofilefraction=100 -test.timeout 25s 2>&1
```

Before:
<img width="1906" alt="Screenshot 2025-05-12 at 3 40 08 PM" src="https://github.com/user-attachments/assets/0059db5d-fc17-4344-ae1e-7d37770fba0c" />

After:
<img width="1908" alt="Screenshot 2025-05-12 at 3 39 48 PM" src="https://github.com/user-attachments/assets/f3ee3869-8b46-45bc-bf63-ba2af958cb84" />


References: #140235

Release note: None

iskettaneh committed:
kvserver: reduce replica mu contention in updateProposalQuotaRaftMuLocked

This commit refactors updateProposalQuotaRaftMuLocked() to reduce the
replica mutex contention, especially in the normal case (QuotaPool is
disabled, and the leadership is stable).

Changes done by this commit:

1. Create a short path for the case where we are not the leader.

2. Create a short path for the case where QuotaPool is not enabled.
We only take the write lock in cases where we already had proposal quota
that needs to be release, or if the flow control level causes the
replicaFlowControlIntegration.onRaftTicked to be a no-op.

3. Untangle the rest of the cases where the QuotaPool is enabled.

References: #140235

Release note: None

iskettaneh committed:
kvserver: remove resetting LastReplicaAdded outside UpdateProposalQuota

This commit removes the logic of resetting LastReplicaAdded from
updateProposalQuotaRaftMuLocked since it has nothing to do with proposal
quota. This should open the door to refactoring
updateProposalQuotaRaftMuLocked and reducing the replica mutex
contention in it.

References: #140235

Release note: None

iskettaneh committed:
kvserver: remove getQuotaPoolEnabled dependency on Replica mutex

This commit removes the need of replica mutex being held in read mode
for (*Replica).getQuotaPoolEnabled(). It does so by making the method
take the range descriptor directly.

References: #140235

Release note: None

iskettaneh committed:
kvserver: remove lastUpdateTimes initialization or reset from updateProposalQuota

This commit removes the logic that initializes/resets the
replica's lastUpdateTimes map outside
updateProposalQuotaRaftMuLocked(). This should open the door
to refactoring updateProposalQuotaRaftMuLocked() to reduce the
replica mutex.

References: #140235

Release note: None

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146561)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`pkg/kv/kvserver/replica_proposal_quota.go` line 190 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/146561#-OQ58j8CCfFTBw7pxuS-:-OQ58j8CCfFTBw7pxuS0:b-sim1aw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae2fbaa08cbb8c64dd2491d011a0013766097bf1/pkg/kv/kvserver/replica_proposal_quota.go#L190)):*
> ```Go
> 			// are not using flowControlV2. If we are using flowControlV2, the calls
> 			// or no-ops.
> 			r.mu.Lock()
> ```

@kvoli Is it safe to assume that if r.raftMu.flowControlLevel > kvflowcontrol.V2NotEnabledWhenLeader, the r.mu.replicaFlowControlIntegration.onRaftTicked() is basically a no-op and we don't really need to hold the replica mutex?

Reading the code seems to indicate that this is the case:
1) On replica init, we set it to no-op if it's > rkvflowcontrol.V2NotEnabledWhenLeader
2) At the beginning of handleRaftReadyRaftMuLocked, we set it to no-op if it's > kvflowcontrol.V2NotEnabledWhenLeader


<!-- Sent from Reviewable.io -->

tbg commented on pkg/kv/kvserver/replica_raft.go:
```diff
@@ -1041,6 +1041,7 @@ func (r *Replica) handleRaftReadyRaftMuLocked(
 		// leader replica, no Ready is emitted. But given that the third
 		// replica has caught up, we can release
 		// some quota back to the pool.
+		r.maybeInitOrResetLastUpdateTimes(lastLeaderID, r.shMu.leaderID /* currentLeaderId */)
```
do we need this here? I'm inclined to think that we don't.

tbg reviewed (approved):
Strange that we're not seeing a benefit in the benchmark. I'll take another look at that as well.

kvoli reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146561)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh, @kvoli, and @tbg)

___
*[`pkg/kv/kvserver/replica_proposal_quota.go` line 190 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/146561#-OQ58j8CCfFTBw7pxuS-:-OQ5CwXXA37I1r3_lN95:bdy767i) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae2fbaa08cbb8c64dd2491d011a0013766097bf1/pkg/kv/kvserver/replica_proposal_quota.go#L190)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

@kvoli Is it safe to assume that if r.raftMu.flowControlLevel > kvflowcontrol.V2NotEnabledWhenLeader, the r.mu.replicaFlowControlIntegration.onRaftTicked() is basically a no-op and we don't really need to hold the replica mutex?

Reading the code seems to indicate that this is the case:
1) On replica init, we set it to no-op if it's > rkvflowcontrol.V2NotEnabledWhenLeader
2) At the beginning of handleRaftReadyRaftMuLocked, we set it to no-op if it's > kvflowcontrol.V2NotEnabledWhenLeader
</blockquote></details>

That is right. Once its above `kvflowcontrol.V2NotEnabledWhenLeader` we call into the rac2 processor code path instead.


<!-- Sent from Reviewable.io -->

iskettaneh committed:
kvserver: clean-up updateProposalQuotaRaftMuLocked

This commit does a couple of cleanups in the method
(*Replica).updateProposalQuotaRaftMuLocked(). The cleanups are mainly
intended to make it more clear and readable.

References: #140235

Release note: None

iskettaneh commented:
TFTR!

bors r+
